exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java

```diff
@@ -130,562 +145,248 @@ public IterOutcome innerNext() {
   }
 
   @Override
-  public WritableBatch getWritableBatch() {
-    return WritableBatch.get(this);
+  public int getRecordCount() {
+    return recordCount;
   }
 
-  private void setValueCount(int count) {
-    for (ValueVector v : allocationVectors) {
-      ValueVector.Mutator m = v.getMutator();
-      m.setValueCount(count);
-    }
-  }
 
-  private boolean doAlloc() {
-    for (ValueVector v : allocationVectors) {
-      try {
-        AllocationHelper.allocateNew(v, current.getRecordCount());
-      } catch (OutOfMemoryException ex) {
-        return false;
-      }
+  @SuppressWarnings("resource")
+  private IterOutcome doWork(RecordBatch inputBatch, boolean newSchema) throws ClassTransformationException, IOException, SchemaChangeException {
+    if (inputBatch.getSchema().getFieldCount() != container.getSchema().getFieldCount()) {
+      // wrong.
     }
-    return true;
-  }
 
-  @SuppressWarnings("resource")
-  private IterOutcome doWork() throws ClassTransformationException, IOException, SchemaChangeException {
-    if (allocationVectors != null) {
-      for (ValueVector v : allocationVectors) {
-        v.clear();
-      }
+    if (newSchema) {
+      createUnionAller(inputBatch);
     }
 
-    allocationVectors = Lists.newArrayList();
-    transfers.clear();
+    container.zeroVectors();
 
-    // If both sides of Union-All are empty
-    if(unionAllInput.isBothSideEmpty()) {
-      for(int i = 0; i < outputFields.size(); ++i) {
-        final String colName = outputFields.get(i).getPath();
-        final MajorType majorType = MajorType.newBuilder()
-            .setMinorType(MinorType.INT)
-            .setMode(DataMode.OPTIONAL)
-            .build();
-
-        MaterializedField outputField = MaterializedField.create(colName, majorType);
-        ValueVector vv = container.addOrGet(outputField, callBack);
-        allocationVectors.add(vv);
-      }
+    for (final ValueVector v : this.allocationVectors) {
+      AllocationHelper.allocateNew(v, inputBatch.getRecordCount());
+    }
```

Comment:
This is a common operation. Maybe move it into `VectorAccessibleUtilities` or some such. The difference here is that `allocationVectors` (no need for `this.`) is a list of vectors, not an iterator over vector wrappers.

Reply:
Sounds reasonable. I put two util method in `VectorUtil`, since `VectorAccessibleUtilities` probably should only take `VectorAccessible` related util method, as its comment stated.